bitcoin/src/crypto/sighash.rs

```diff
@@ -64,24 +66,704 @@ sha256t_hash_newtype! {
     pub struct TapSighash(_);
 }
 
+/// Represents types containing transaction data.
+///
+/// This is most often `Transaction` but types that contain transaction data but have different
+/// layout may implement this to allow signing without having to construct `Transaction`.
+pub trait AsTransaction {
+    /// Returns the number of inputs in the transaction.
+    fn inputs_len(&self) -> usize;
+
+    /// Returns an input at given `index`.
+    ///
+    /// Returns `None` if `index > self.inputs_len()`.
+    fn input_at(&self, index: usize) -> Option<&TxIn>;
+
+    /// Returns the number of outputs in the transaction.
+    fn outputs_len(&self) -> usize;
+
+    /// Returns an output at given `index`.
+    ///
+    /// Returns `None` if `index > self.outputs_len()`.
+    fn output_at(&self, index: usize) -> Option<&TxOut>;
+
+    /// Returns transaction version.
+    fn version(&self) -> transaction::Version;
+
+    /// Returns transaction lock time.
+    fn lock_time(&self) -> absolute::LockTime;
+}
+
+/// Represents mutably-acessible types containing transaction data.
+pub trait AsMutTransaction: AsTransaction {
+    /// Returns a mutable reference to the input at given `index`.
+    ///
+    /// Returns `None` if `index > self.inputs_len()`.
+    fn mut_input_at(&mut self, index: usize) -> Option<&mut TxIn>;
+}
+
+/// Represents types containing transaction data with extended inputs.
+pub trait AsExtendedTransaction: AsTransaction {
```

Comment:
What is the purpose of having this separate to `AsTransaction`? I was under the impression the the `Ext` idiom was for adding functionality to a trait from a different crate.

Reply:
`Transaction` doesn't contain prevouts and it seems reasonable to continue to supporting the existing API (modulo duplicated index).

`Ext` is for extension traits, which are traits that are implemented for a single specific type or a trait.

Also the wording is in line with `AsRef`/`AsMut`.

Reply:
Ugh, I commented on the wrong line. I meant to comment on `trait AsTransactionExt: AsTransaction {`.

Reply:
```rust
trait AsTransactionExt: AsTransaction {
    fn inputs(&self) -> Inputs<'_, Self> {
        Inputs(self)
    }

    fn outputs(&self) -> Outputs<'_, Self> {
        Outputs(self)
    }
}
```
This code could be in `AsTransaction` I was wondering why it was a separate trait.